framework/src/utils/SlepcSupport.C

```diff
@@ -131,56 +131,82 @@ setSlepcEigenSolverTolerances(EigenProblem & eigen_problem,
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "snes_max_it",
+        stringify(params.get<unsigned int>("nl_max_its")));
+
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "snes_max_funcs",
+        stringify(params.get<unsigned int>("nl_max_funcs")));
+
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "snes_atol",
+        stringify(params.get<Real>("nl_abs_tol")));
+
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "snes_rtol",
+        stringify(params.get<Real>("nl_rel_tol")));
+
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "snes_stol",
+        stringify(params.get<Real>("nl_rel_step_tol")));
 
     // linear solver
-    Moose::PetscSupport::setSinglePetscOption(solver_params._prefix + "ksp_max_it",
-                                              stringify(params.get<unsigned int>("l_max_its")));
-
-    Moose::PetscSupport::setSinglePetscOption(solver_params._prefix + "ksp_rtol",
-                                              stringify(params.get<Real>("l_tol")));
-
-    Moose::PetscSupport::setSinglePetscOption(solver_params._prefix + "ksp_atol",
-                                              stringify(params.get<Real>("l_abs_tol")));
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "ksp_max_it",
+        stringify(params.get<unsigned int>("l_max_its")));
+
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(dont_add_these_options,
+                                                           solver_params._prefix + "ksp_rtol",
+                                                           stringify(params.get<Real>("l_tol")));
+
+    Moose::PetscSupport::setSinglePetscOptionIfAppropriate(
+        dont_add_these_options,
+        solver_params._prefix + "ksp_atol",
+        stringify(params.get<Real>("l_abs_tol")));
   }
   else
   { // linear eigenvalue problem
     // linear solver
-    Moose::PetscSupport::setSinglePetscOption(solver_params._prefix + "st_ksp_max_it",
-                                              stringify(params.get<unsigned int>("l_max_its")));
-
-    Moose::PetscSupport::setSinglePetscOption(solver_params._prefix + "st_ksp_rtol",
-                                              stringify(params.get<Real>("l_tol")));
-
-    Moose::PetscSupport::setSinglePetscOption(solver_params._prefix + "st_ksp_atol",
-                                              stringify(params.get<Real>("l_abs_tol")));
+    const auto & dont_add_these_options = eigen_problem.getPetscOptions().dont_add_these_options;
```

Comment:
Isn't this shadowing the previous `dont_add_these_options`?

Reply:
Good catch, I'll take it out.